muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java

```diff
@@ -167,9 +178,105 @@ public Response listWorkspaceActivities(
         workspaceIdentifier,
         includeTransferCredits,
         includeAssignmentStatistics);
+    
+    Integer allCourseCredits = 0;
+    Integer mandatoryCourseCredits = 0;
+    
+    for (WorkspaceActivity activity : activityInfo.getActivities()) {
+      
+      for (WorkspaceActivitySubject workspaceActivitySubject : activity.getSubjects()) {
+
+        List<WorkspaceAssessmentState> assessmentStatesList = activity.getAssessmentStates();
+        Integer size = assessmentStatesList.size() - 1;
+        
+        if (!assessmentStatesList.isEmpty()) {
+          WorkspaceAssessmentState assessmentState = assessmentStatesList.get(size);
+          if (assessmentState.getState() == WorkspaceAssessmentState.PASS || assessmentState.getState() == WorkspaceAssessmentState.TRANSFERRED) {
+            if (workspaceActivitySubject.getCourseLengthSymbol().equals("op")) {
+              
+              for (WorkspaceActivityCurriculum curriculum : activity.getCurriculums()) {
+                if (curriculum.getName().equals("OPS 2021")) {
+                  int units = workspaceActivitySubject.getCourseLength().intValue();
+                  
+                  // All completed courses
+                  allCourseCredits = Integer.sum(units, allCourseCredits);
+                  
+                  
+                  // Mandatority for transferred courses
+                  // Transferred courses doesn't have ids or identifiers so that's why these need to get separately
+                  if (activity.getId() == null && assessmentState.getState() == WorkspaceAssessmentState.TRANSFERRED) {
+                    Mandatority mandatority = activity.getMandatority();
+                    if (mandatority != null && mandatority == Mandatority.MANDATORY) {
+                      mandatoryCourseCredits = Integer.sum(units, mandatoryCourseCredits);
+                   }
+                  }
+                  
+                  // Search for finding out course mandaority
+                  SearchProvider searchProvider = getProvider("elastic-search");
```

Comment:
Move this outside all loops, it always resolves to same result and is not depending on any loop variables.